File: core/causal_intelligence/semantic_causal_intelligence_orchestrator.py
from __future__ import annotations

from typing import Any, Dict

from .semantic_causality_graph_engine import (
    build_semantic_causality_graph,
)
from .runtime_failure_lineage_engine import (
    build_runtime_failure_lineage,
)
from .distributed_propagation_engine import (
    propagate_distributed_state,
)
from .semantic_recovery_causality_engine import (
    analyze_recovery_causality,
)
from .runtime_equilibrium_engine import (
    compute_runtime_equilibrium,
)
from .semantic_instability_forecast_engine import (
    forecast_runtime_instability,
)
from .execution_timing_semantics_engine import (
    analyze_execution_timing,
)
from .dependency_cascade_intelligence_engine import (
    analyze_dependency_cascade,
)
from .semantic_runtime_mutation_lineage_engine import (
    build_mutation_lineage,
)
from .distributed_scheduling_pressure_engine import (
    compute_scheduling_pressure,
)
from .runtime_drift_causality_engine import (
    analyze_drift_causality,
)
from .semantic_causal_replay_engine import (
    replay_causal_sequence,
)
from .semantic_recovery_forecast_engine import (
    forecast_recovery_outcome,
)
from .runtime_semantic_equilibrium_engine import (
    assess_semantic_equilibrium,
)
from .semantic_execution_mutation_engine import (
    trace_execution_mutations,
)
from .semantic_runtime_wave_propagation_engine import (
    propagate_runtime_waves,
)
from .distributed_runtime_causal_graph_engine import (
    build_distributed_causal_graph,
)
from .semantic_runtime_resonance_engine import (
    measure_runtime_resonance,
)
from .semantic_execution_stability_horizon_engine import (
    forecast_stability_horizon,
)


def orchestrate_semantic_causal_intelligence(
    runtime_ir: Dict[str, Any],
) -> Dict[str, Any]:

    causality_graph = build_semantic_causality_graph(
        runtime_ir
    )
    runtime_ir["runtime_causality_graph"] = causality_graph

    failure_lineage = build_runtime_failure_lineage(
        runtime_ir
    )
    propagation = propagate_distributed_state(
        runtime_ir
    )
    runtime_ir["distributed_propagation"] = propagation

    recovery = analyze_recovery_causality(runtime_ir)
    runtime_ir["recovery_causality"] = recovery

    equilibrium = compute_runtime_equilibrium(
        runtime_ir
    )
    runtime_ir["runtime_equilibrium"] = equilibrium

    instability = forecast_runtime_instability(
        runtime_ir
    )
    runtime_ir["instability_forecast"] = instability

    timing = analyze_execution_timing(runtime_ir)
    cascade = analyze_dependency_cascade(runtime_ir)
    mutation_lineage = build_mutation_lineage(runtime_ir)
    scheduling_pressure = compute_scheduling_pressure(
        runtime_ir
    )
    baseline = dict(runtime_ir.get("baseline", {}) or {})
    drift_causality = analyze_drift_causality(
        runtime_ir, baseline
    )
    causal_replay = replay_causal_sequence(runtime_ir)
    recovery_forecast = forecast_recovery_outcome(
        runtime_ir
    )
    semantic_equilibrium = assess_semantic_equilibrium(
        runtime_ir
    )
    execution_mutations = trace_execution_mutations(
        list(runtime_ir.get("transitions", []) or [])
    )
    waves = propagate_runtime_waves(
        propagation.get("propagation_paths", [])
    )
    distributed_causal_graph = build_distributed_causal_graph(
        runtime_ir
    )
    resonance = measure_runtime_resonance(runtime_ir)
    stability_horizon = forecast_stability_horizon(
        runtime_ir
    )

    return {
        "causality_graph": causality_graph,
        "failure_lineage": failure_lineage,
        "distributed_propagation": propagation,
        "recovery_causality": recovery,
        "runtime_equilibrium": equilibrium,
        "instability_forecast": instability,
        "execution_timing": timing,
        "dependency_cascade": cascade,
        "mutation_lineage": mutation_lineage,
        "scheduling_pressure": scheduling_pressure,
        "drift_causality": drift_causality,
        "causal_replay": causal_replay,
        "recovery_forecast": recovery_forecast,
        "semantic_equilibrium": semantic_equilibrium,
        "execution_mutations": execution_mutations,
        "wave_propagation": waves,
        "distributed_causal_graph": distributed_causal_graph,
        "runtime_resonance": resonance,
        "stability_horizon": stability_horizon,
        "deterministic": True,
        "bounded": True,
    }

File: src/causal_intelligence/semanticCausalIntelligenceOrchestrator.ts
/**
 * Converted from Python: core/causal_intelligence/semantic_causal_intelligence_orchestrator.py
 * @generated — WebWeaveX python→javascript library port
 */

import * as py from "../runtime/pyCompat.js";
import { buildSemanticCausalityGraph } from "./semanticCausalityGraphEngine.js";
import { buildRuntimeFailureLineage } from "./runtimeFailureLineageEngine.js";
import { propagateDistributedState } from "./distributedPropagationEngine.js";
import { analyzeRecoveryCausality } from "./semanticRecoveryCausalityEngine.js";
import { computeRuntimeEquilibrium } from "./runtimeEquilibriumEngine.js";
import { forecastRuntimeInstability } from "./semanticInstabilityForecastEngine.js";
import { analyzeExecutionTiming } from "./executionTimingSemanticsEngine.js";
import { analyzeDependencyCascade } from "./dependencyCascadeIntelligenceEngine.js";
import { buildMutationLineage } from "./semanticRuntimeMutationLineageEngine.js";
import { computeSchedulingPressure } from "./distributedSchedulingPressureEngine.js";
import { analyzeDriftCausality } from "./runtimeDriftCausalityEngine.js";
import { replayCausalSequence } from "./semanticCausalReplayEngine.js";
import { forecastRecoveryOutcome } from "./semanticRecoveryForecastEngine.js";
import { assessSemanticEquilibrium } from "./runtimeSemanticEquilibriumEngine.js";
import { traceExecutionMutations } from "./semanticExecutionMutationEngine.js";
import { propagateRuntimeWaves } from "./semanticRuntimeWavePropagationEngine.js";
import { buildDistributedCausalGraph } from "./distributedRuntimeCausalGraphEngine.js";
import { measureRuntimeResonance } from "./semanticRuntimeResonanceEngine.js";
import { forecastStabilityHorizon } from "./semanticExecutionStabilityHorizonEngine.js";

export function orchestrateSemanticCausalIntelligence(runtime_ir: any): any {
  var causality_graph: any = buildSemanticCausalityGraph(runtime_ir);
  py.setItem(runtime_ir, "runtime_causality_graph", causality_graph);
  var failure_lineage: any = buildRuntimeFailureLineage(runtime_ir);
  var propagation: any = propagateDistributedState(runtime_ir);
  py.setItem(runtime_ir, "distributed_propagation", propagation);
  var recovery: any = analyzeRecoveryCausality(runtime_ir);
  py.setItem(runtime_ir, "recovery_causality", recovery);
  var equilibrium: any = computeRuntimeEquilibrium(runtime_ir);
  py.setItem(runtime_ir, "runtime_equilibrium", equilibrium);
  var instability: any = forecastRuntimeInstability(runtime_ir);
  py.setItem(runtime_ir, "instability_forecast", instability);
  var timing: any = analyzeExecutionTiming(runtime_ir);
  var cascade: any = analyzeDependencyCascade(runtime_ir);
  var mutation_lineage: any = buildMutationLineage(runtime_ir);
  var scheduling_pressure: any = computeSchedulingPressure(runtime_ir);
  var baseline: any = py.pyDict(py.or2(py.get(runtime_ir, "baseline", {}), () => ({})));
  var drift_causality: any = analyzeDriftCausality(runtime_ir, baseline);
  var causal_replay: any = replayCausalSequence(runtime_ir);
  var recovery_forecast: any = forecastRecoveryOutcome(runtime_ir);
  var semantic_equilibrium: any = assessSemanticEquilibrium(runtime_ir);
  var execution_mutations: any = traceExecutionMutations([...py.iter(py.or2(py.get(runtime_ir, "transitions", []), () => ([])))]);
  var waves: any = propagateRuntimeWaves(py.get(propagation, "propagation_paths", []));
  var distributed_causal_graph: any = buildDistributedCausalGraph(runtime_ir);
  var resonance: any = measureRuntimeResonance(runtime_ir);
  var stability_horizon: any = forecastStabilityHorizon(runtime_ir);
  return {"causality_graph": causality_graph, "failure_lineage": failure_lineage, "distributed_propagation": propagation, "recovery_causality": recovery, "runtime_equilibrium": equilibrium, "instability_forecast": instability, "execution_timing": timing, "dependency_cascade": cascade, "mutation_lineage": mutation_lineage, "scheduling_pressure": scheduling_pressure, "drift_causality": drift_causality, "causal_replay": causal_replay, "recovery_forecast": recovery_forecast, "semantic_equilibrium": semantic_equilibrium, "execution_mutations": execution_mutations, "wave_propagation": waves, "distributed_causal_graph": distributed_causal_graph, "runtime_resonance": resonance, "stability_horizon": stability_horizon, "deterministic": true, "bounded": true};
}
export { analyzeDependencyCascade, analyzeDriftCausality, analyzeExecutionTiming, analyzeRecoveryCausality, assessSemanticEquilibrium, buildDistributedCausalGraph, buildMutationLineage, buildRuntimeFailureLineage, buildSemanticCausalityGraph, computeRuntimeEquilibrium, computeSchedulingPressure, forecastRecoveryOutcome, forecastRuntimeInstability, forecastStabilityHorizon, measureRuntimeResonance, propagateDistributedState, propagateRuntimeWaves, replayCausalSequence, traceExecutionMutations };
